### DISCUSSION TITLE
Setup User Authentication with Lucia

### DISCUSSION THREAD

**Discussion by NathanGrenier** [NathanGrenier/SOEN-341#32] [Ideas]
I recommend using [Lucia](https://lucia-auth.com) for user authentication. If you would rather use another library, feel free to comment on this discussion.

**From old issue:**
> We will be using [Lucia](https://lucia-auth.com) for user authentication. Lucia also provides a [prisma adapter](https://lucia-auth.com/database-adapters/prisma/).
> 
> [Lucia SvelteKit quickstart guide](https://lucia-auth.com/getting-started/sveltekit/)
> 
> Steps:
> 1. Follow the setup guide in [this](https://www.youtube.com/watch?v=UMpKaZy0Rpc) video.
> 2. Setup a login, logout, and register page.
> 3. Integrate OAuth (probably Google and/or Github).

**Comment by briantkatch** [comment]
Do we need to introduce a dependency for authentication?

Most likely this project is going to be username and passsword (this is what past course outlines seem to call for), no OAuth, passkeys or anything fancy. 

The Lucia project doesn't seem to have any commercial backing to it from what I can see on their website. And given the risks involved with dependencies and their own dependencies (left-pad, [security](https://david-gilbertson.medium.com/im-harvesting-credit-card-numbers-and-passwords-from-your-site-here-s-how-9a8cb347c5b5)) it's something to consider.

**Comment by NathanGrenier** [comment]
> Do we need to introduce a dependency for authentication?

No, but it generally makes things much easier to implement.

> Most likely this project is going to be username and password (this is what past course outlines seem to call for), no OAuth, passkeys or anything fancy.

I don't think they'll be super strict about needing specifically a username and password, just a way to register / login. What I was planning to do is rely only on OAuth so that we don't have to store any passwords at all. If it turns out that we do need to store user's passwords, Lucia provides a way for a user to have multiple authentication methods (Google OAuth and the traditional username and password).

> The Lucia project doesn't seem to have any commercial backing to it from what I can see on their website.

Not too sure why commercial backing is super important. There is a chance that Lucia stop being maintained and that causes breaking changes, but the scope of this project is just a hobby/school project. If that's too big of an issue / risk, we could decide to use another SvelteKit compatible authentication library that has more longevity / backing.

> And given the risks involved with dependencies and their own dependencies (left-pad, [security](https://david-gilbertson.medium.com/im-harvesting-credit-card-numbers-and-passwords-from-your-site-here-s-how-9a8cb347c5b5)) it's something to consider.

Adding Lucia to the project would add another dependency, but I don't see why that's a problem if it makes the authentication process easier to implement. I doesn't seem like Lucia relies on any [dependencies itself](https://www.npmjs.com/package/lucia?activeTab=dependencies), so it would only add 1 more to the project.

**Reply by NathanGrenier** [reply]
The more popular authentication library is [Auth.js](https://authjs.dev/). This was the other option I was going to look at before implementing authentication.

**Comment by briantkatch** [comment]
I think we'll probably have to have usernames and passwords as an option because that's the easiest thing to hand over to a TA to mark the project (not every TA is going to have a Google account they are comfortable giving to some student project. Also Google gives a scary warning when OAuthing to an app until you go through a process to get 'verified' with Google (it's not terribly difficult, I've done it, but requires a public app with a documented privacy policy, etc.))

None of the libraries linked support [Secure Remote Password](https://datatracker.ietf.org/doc/html/rfc2945) anyway, which is the most secure way to do password-based authentication from a cryptographic standpoint (the linked libraries require sending the password in raw form over TLS, rather than using a challenge-response mechanism to arrive at a shared secret). 

If we really need to bring on a dependency, I'd say the risks presented by the use of Lucia are manageable in this context. It might be preferable to Auth.js as it has no child dependencies and I'm not sure at first glance how mature Auth.js's integration with Svelte is - there seem to be some issues related to it in the issue tracker.

**Reply by NathanGrenier** [reply]
> I think we'll probably have to have usernames and passwords as an option because that's the easiest thing to hand over to a TA to mark the project 

You're probably right. I was still planning on using a library to manage the users and their sessions in the database.

> (not every TA is going to have a Google account they are comfortable giving to some student project.

Almost everyone has a google account and I don't think they'll care about using it to register. Most people trust the OAuth when they see it. Either way, letting them register with a username and password would be a good default option to include.

> Also Google gives a scary warning when OAuthing to an app until you go through a process to get 'verified' with Google (it's not > terribly difficult, I've done it, but requires a public app with a documented privacy policy, etc.))

Didn't know about this but based on what you said, it doesn't seem too hard to verify.

> None of the libraries linked support [Secure Remote Password](https://datatracker.ietf.org/doc/html/rfc2945) anyway, which is the most secure way to do password-based authentication from a cryptographic standpoint (the linked libraries require sending the password in raw form over TLS, rather than using a challenge-response mechanism to arrive at a shared secret).

Not too sure about this standard you're talking about. I would trust Lucia and Auth.js security implementation more than the one I implement myself. This is just a hobby / school level project, I'm not interested in spending a lot of time making sure authentication is super secure. That's a why I wanted to offload the development effort by just using an existing library.

> I'm not sure at first glance how mature Auth.js's integration with Svelte is - there seem to be some issues related to it in the issue tracker.

Ya, that's why I leaned towards Lucia first. When I focus more on the authentication part of the app I'll take a closer look at Lucia and Auth.js and see if they're worth using. 

I think that having an api that's easy to setup and that you can use to get user's authentication credentials is worthwhile. I'd trust that both libraries are secure enough to use and that anything more is overkill for the scope of this project.
